File: backend/schemas/stamp.py
from datetime import date, datetime
from typing import Optional
from pydantic import BaseModel


# Prefecture Stamp Schemas
class PrefectureStampBase(BaseModel):
    prefecture_code: str
    name: str
    name_romaji: str
    image_url: str
    region: str
    display_order: int
    is_active: bool = True


class PrefectureStampResponse(PrefectureStampBase):
    created_at: datetime

    class Config:
        from_attributes = True


# User Stamp Collection Schemas
class StampCollectionSummary(BaseModel):
    """ユーザーのスタンプ収集サマリー"""

    total_prefectures: int  # 訪問都道府県数
    total_visits: int  # 総訪問回数
    total_farms: int  # 訪問ファーム数
    completion_rate: float  # 達成率 (0-100)


class PrefectureStampStatus(BaseModel):
    """都道府県別スタンプ状態"""

    prefecture_code: str
    name: str
    image_url: str
    region: str
    is_visited: bool
    visit_count: int = 0
    first_visit_date: Optional[date] = None
    last_visit_date: Optional[date] = None
    unique_farms_count: int = 0


class StampCollectionResponse(BaseModel):
    """スタンプコレクション全体のレスポンス"""

    summary: StampCollectionSummary
    stamps: list[PrefectureStampStatus]

    class Config:
        from_attributes = True


# Prefecture Detail Schemas
class VisitedFarmInfo(BaseModel):
    """訪問済みファーム情報"""

    farm_id: int
    farm_name: str
    visit_date: date
    experience_type: str
    review_id: int


class PrefectureDetailResponse(BaseModel):
    """特定都道府県の詳細情報"""

    prefecture_code: str
    name: str
    visit_count: int
    first_visit_date: date
    last_visit_date: date
    unique_farms_count: int
    visited_farms: list[VisitedFarmInfo]

    class Config:
        from_attributes = True

File: frontend/src/types/stamp.ts
/**
 * Stamp Rally Type Definitions
 */

export interface PrefectureStamp {
  prefecture_code: string;
  name: string;
  name_romaji: string;
  image_url: string;
  region: string;
  display_order: number;
  is_active: boolean;
  created_at: string;
}

export interface StampCollectionSummary {
  total_prefectures: number; // 訪問都道府県数
  total_visits: number; // 総訪問回数
  total_farms: number; // 訪問ファーム数
  completion_rate: number; // 達成率 (0-100)
}

export interface PrefectureStampStatus {
  prefecture_code: string;
  name: string;
  image_url: string;
  region: string;
  is_visited: boolean;
  visit_count: number;
  first_visit_date: string | null;
  last_visit_date: string | null;
  unique_farms_count: number;
}

export interface StampCollectionResponse {
  summary: StampCollectionSummary;
  stamps: PrefectureStampStatus[];
}

export interface VisitedFarmInfo {
  farm_id: number;
  farm_name: string;
  visit_date: string;
  experience_type: string;
  review_id: number;
}

export interface PrefectureDetailResponse {
  prefecture_code: string;
  name: string;
  visit_count: number;
  first_visit_date: string;
  last_visit_date: string;
  unique_farms_count: number;
  visited_farms: VisitedFarmInfo[];
}

// 地方区分マッピング
export const REGION_NAMES: Record<string, string> = {
  hokkaido: "北海道",
  tohoku: "東北",
  kanto: "関東",
  chubu: "中部",
  kinki: "近畿",
  chugoku: "中国",
  shikoku: "四国",
  kyushu: "九州・沖縄",
};
